java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -256,8 +268,21 @@ private static boolean isVolatileField(Symbol symbol) {
     }
   }
 
-  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Object constraint) {
-    return SymbolicValue.isDisposable(symbolicValue) && (constraint == null || !(constraint instanceof ObjectConstraint) || ((ObjectConstraint) constraint).isDisposable());
+  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Constraint constraint) {
+    //FIXME this should be handle with callbacks rather than keeping those value in programstate
+    return SymbolicValue.isDisposable(symbolicValue) &&
+      (constraint == null ||
+        !(constraint instanceof UnclosedResourcesCheck.ResourceConstraint
```

Comment:
Why only these 3 checks? I can't get the logic saying that we have to keep constraints from these checks.